masonry/src/widget/grid.rs

```diff
@@ -0,0 +1,442 @@
+
+    fn on_status_change(&mut self, _ctx: &mut LifeCycleCtx, _event: &StatusChange) {}
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle) {
+        for child in self.children.iter_mut().filter_map(|x| x.widget_mut()) {
+            child.lifecycle(ctx, event);
+        }
+    }
+
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
+        bc.debug_check("Grid");
+        /*let bc_changed = self.old_bc != *bc;
+        if bc_changed {
+            self.old_bc = *bc;
+            if !self.needs_layout {
+                self.needs_layout = true;
+            }
+        }*/
+        let total_size = bc.max();
+        let width_unit = (total_size.width + self.grid_spacing) / (self.grid_width as f64);
+        let height_unit = (total_size.height + self.grid_spacing) / (self.grid_height as f64);
+        for child in &mut self.children {
+            /*if !self.needs_layout && !ctx.child_needs_layout(&child.widget) {
+                ctx.mark_child_as_visited(&child.widget, true);
+                continue; // TODO: This breaks it. This is an attempted optimization.
+            }*/
+            let cell_size = Size::new(
+                child.width as f64 * width_unit - self.grid_spacing,
+                child.height as f64 * height_unit - self.grid_spacing,
+            );
+            let child_bc = BoxConstraints::new(cell_size, cell_size);
+            let _ = child.widget.layout(ctx, &child_bc);
+            ctx.place_child(
+                &mut child.widget,
+                Point::new(child.x as f64 * width_unit, child.y as f64 * height_unit),
+            );
+        }
+        /*if self.needs_layout {
+            self.needs_layout = false;
+        }*/
+        total_size
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
+        // paint the baseline if we're debugging layout
+        if ctx.debug_paint && ctx.widget_state.baseline_offset != 0.0 {
+            let color = get_debug_color(ctx.widget_id().to_raw());
+            let my_baseline = ctx.size().height - ctx.widget_state.baseline_offset;
+            let line = Line::new((0.0, my_baseline), (ctx.size().width, my_baseline));
+
+            let stroke_style = Stroke::new(1.0).with_dashes(0., [4.0, 4.0]);
+            scene.stroke(&stroke_style, Affine::IDENTITY, color, None, &line);
+        }
+    }
+
+    fn accessibility_role(&self) -> Role {
+        Role::GenericContainer
+    }
+
+    fn accessibility(&mut self, _: &mut AccessCtx) {}
+
+    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
+        self.children
+            .iter()
+            .filter_map(|child| child.widget())
+            .map(|widget_pod| widget_pod.id())
+            .collect()
+    }
+
+    fn make_trace_span(&self) -> Span {
+        trace_span!("Grid")
+    }
+}
+
+struct Child {
+    widget: WidgetPod<Box<dyn Widget>>,
+    x: i32,
+    y: i32,
+    width: i32,
+    height: i32,
+}
```

Comment:
I'd rather follow the same order as the rest of the widget files:

- First define the types.
- Then inherent impls for these types.
- Then WidgetMut.
- Then impl Widget.
- Then tests.

Flex doesn't follow that order, but I think it's the only one.

Reply:
I changed the order to match what you're requesting. There are several types, so let me know if I didn't handle the ordering of that stuff the way you wanted.

Reply:
It's all good. The marker comments are a nice touch.